handlers/product-switch-api/src/frequencySwitchEndpoint.ts

```diff
@@ -0,0 +1,693 @@
+ * @throws ValidationError if subscription fails any validation checks
+ */
+export async function selectCandidateSubscriptionCharge(
+	subscription: ZuoraSubscription,
+	today: dayjs.Dayjs,
+	account: ZuoraAccount,
+	productCatalog: ProductCatalog,
+	zuoraClient: ZuoraClient,
+): Promise<{ ratePlan: RatePlan; charge: RatePlanCharge }> {
+	const eligibilityChecker = new EligibilityChecker();
+
+	await eligibilityChecker.assertGenerallyEligible(
+		subscription,
+		account.metrics.totalInvoiceBalance,
+		async () => {
+			const billingPreview = await getBillingPreview(
+				zuoraClient,
+				today.add(13, 'months'),
+				subscription.accountNumber,
+			);
+			const invoiceItems = toSimpleInvoiceItems(
+				itemsForSubscription(subscription.subscriptionNumber)(billingPreview),
+			);
+			return getNextInvoiceItems(invoiceItems);
+		},
+	);
+
+	// Find the SupporterPlus Monthly rate plan using the catalog
+	const supporterPlusMonthlyRatePlanId =
+		productCatalog.SupporterPlus.ratePlans.Monthly.id;
+
+	const supporterPlusMonthlyRatePlan = subscription.ratePlans.find(
+		(ratePlan) =>
+			ratePlan.lastChangeType !== 'Remove' &&
+			supporterPlusMonthlyRatePlanId === ratePlan.productRatePlanId,
+	);
+
+	assertValidState(
+		supporterPlusMonthlyRatePlan !== undefined,
+		frequencySwitchValidationRequirements.hasEligibleCharges,
+		'SupporterPlus Monthly rate plan not found',
+	);
+
+	// Find the subscription charge using the catalog's product rate plan charge ID
+	const subscriptionChargeId =
+		productCatalog.SupporterPlus.ratePlans.Monthly.charges.Subscription.id;
+
+	const subscriptionCharge = supporterPlusMonthlyRatePlan.ratePlanCharges.find(
+		(charge) => charge.productRatePlanChargeId === subscriptionChargeId,
+	);
+
+	assertValidState(
+		subscriptionCharge !== undefined,
+		frequencySwitchValidationRequirements.hasEligibleCharges,
+		`Subscription charge not found in SupporterPlus Monthly rate plan`,
+	);
+
+	// Validate the charge properties
+	assertValidState(
+		subscriptionCharge.type === 'Recurring',
+		frequencySwitchValidationRequirements.hasEligibleCharges,
+		`charge type is "${subscriptionCharge.type}", not "Recurring"`,
+	);
+
+	const todayDate = today.toDate();
+	assertValidState(
+		subscriptionCharge.effectiveStartDate < todayDate,
+		frequencySwitchValidationRequirements.hasEligibleCharges,
+		`effectiveStartDate ${subscriptionCharge.effectiveStartDate.toISOString()} is in the future`,
+	);
+
+	assertValidState(
+		subscriptionCharge.effectiveEndDate >= todayDate,
+		frequencySwitchValidationRequirements.hasEligibleCharges,
+		`effectiveEndDate ${subscriptionCharge.effectiveEndDate.toISOString()} is in the past`,
+	);
+
+	assertValidState(
+		!subscriptionCharge.chargedThroughDate ||
+			subscriptionCharge.chargedThroughDate >= todayDate,
+		frequencySwitchValidationRequirements.hasEligibleCharges,
+		`chargedThroughDate ${subscriptionCharge.chargedThroughDate?.toISOString()} is in the past`,
+	);
```

Comment:
same as "Recurring" above, I think this should always be the case (as we have checked the charge effective dates).  happy to keep it here as a belt and braces check
```suggestion
  // this should always be the case if the bill runs are working, but checking for safety
	assertValidState(
		!subscriptionCharge.chargedThroughDate ||
			subscriptionCharge.chargedThroughDate >= todayDate,
		frequencySwitchValidationRequirements.hasEligibleCharges,
		`chargedThroughDate ${subscriptionCharge.chargedThroughDate?.toISOString()} is in the past`,
	);
```